### PR TITLE
DataFactory: if VDS don't also include individual external datasets

### LINKED ISSUE #285: dxtbx with Diamond nxs files: can access twice as many data images as present on disk
Pointed out by @rjgildea 

80 image grid scan as input:

```
dials.import xtal_3_2.nxs

dials.python -> 

from dxtbx.model import experiment_list
e = experiment_list.ExperimentList.from_file("imported.expt")
i = e[0].imageset
len(i) -> 80
data = i.reader().read(100)
```

_does not fail_ - because this data includes a virtual data set at `/entry/data/data` which is pointing at internal references to `/entry/data/data_000001` which is itself an external reference to another file which _actually_ contains 80 images in this case. But the data are doubled as the VDS and external links are _both_ counted. 

In normal operation this is just funny - as in odd - because we always know the number of images, but it is still something of a sleeper which could catch us out later, in particular with more interesting uses of VDS. 

Proposed fix: _either_ read the VDS _or_ reference the external data and *don't do both*

### PR DESCRIPTION
**Pull request by rjgildea**
Fix bug in `nexus.DataFactory` that allowed access to twice as many images as available on disk for VDS nexus files.

If the node contains a VDS, then only use this dataset.

Fixes #285

### CODE REVIEW

File: format/FormatHDF5EigerNearlyNexus.py [hunk 1]
```diff
@@ -142,6 +142,7 @@ def create_scalar(handle, path, dtype, value):
                 shape=shape,
                 ndim=handle_orig_entry.ndim,
                 filename=handle_orig_entry.file.filename,
+                is_virtual=False,
```
Comment (graeme-winter) [inline format/FormatHDF5EigerNearlyNexus.py:145]: `is_virtual` better than `is_vds`?

_probably_ I would expect the latter

**Review by graeme-winter — APPROVED** [review]
Need to check if there are any performance implications from this change set but overall looks good. 

I have not yet tested but i obviously assume you did before making the PR

As always this raises some questions about the changes in the local area but that is out of context for this PR.

File: format/FormatHDF5EigerNearlyNexus.py (hunk 1 above)
Comment (rjgildea) [reply format/FormatHDF5EigerNearlyNexus.py:145]: This is purely for compatibility with `h5py.Dataset.is_virtual`:
https://docs.h5py.org/en/stable/high/dataset.html#h5py.Dataset.is_virtual

Comment (graeme-winter) [reply format/FormatHDF5EigerNearlyNexus.py:145]: That is good enough for me. I withdraw my comment.

**Comment by rjgildea** [comment]
> Need to check if there are any performance implications from this change set

There shouldn't be any performance implications, as this bit of the code is only hit on initial construction of the `nexus.py` classes, not when extracting every image. If anything it _might_ be slightly faster as the lookup arrays will be half the size for VDS files.

**Comment by ndevenish** [comment]
Stupid question - what's the ` is_virtual=False,` parameter doing?

**Comment by rjgildea** [comment]
> Stupid question - what's the ` is_virtual=False,` parameter doing?

It is an attribute of `h5py.Dataset`
> True if this dataset is a virtual dataset, otherwise False.

https://docs.h5py.org/en/stable/high/dataset.html#h5py.Dataset.is_virtual

Used to control the flow here:
https://github.com/cctbx/dxtbx/pull/286/files#diff-561673f296e5629bbe2e783ce83757f59f54ffe3bc8151f5767a885f78d58ddeR1361-R1365

**Comment by ndevenish** [comment]
Okay, I think I get it, partially it was misreading the diff (looked like a function argument). It's constructing an object that sits in a collection amongst `Dataset` instances. If the `Dataset` is virtual then it replaces the collection of other `DataFactoryCache` objects. 👍

**Pull request merged**